components/match2/wikis/dota2/match_group_input_custom.lua

```diff
@@ -59,648 +44,249 @@ local NOW = os.time(os.date('!*t') --[[@as osdateparam]])
-		if String.startsWith(key, 'match2') then
-			standaloneMatch[key] = nil
-		end
-	end
+	for opponentIndex = 1, opponentCount do
+		local opponentData = {
+			objectives = MatchParser.getObjectives(map, opponentIndex),
+			side = MatchParser.getSide(map, opponentIndex),
+		}
+		opponentData = Table.merge(opponentData,
+			Table.map(MatchParser.getHeroPicks(map, opponentIndex) or {}, function(idx, hero)
+				return 'hero' .. idx, getCharacterName(hero)
+			end),
+			Table.map(MatchParser.getHeroBans(map, opponentIndex) or {}, function(idx, hero)
+				return 'ban' .. idx, getCharacterName(hero)
+			end)
+		)
 
-	-- Copy all match level records which have value
-	for key, value in pairs(standaloneMatch) do
-		if String.isNotEmpty(value) then
-			match[key] = value
-		end
+		Table.mergeInto(extraData, Table.map(opponentData, function(key, value)
+			return prefixKeyWithTeam(key, opponentIndex), value
+		end))
 	end
 
-	return match
-end
-
---
--- map related functions
---
+	extraData.vetophase = MatchParser.getVetoPhase(map)
+	Array.forEach(extraData.vetophase or {}, function(veto)
+		veto.character = getCharacterName(veto.character)
+	end)
 
--- Parse extradata information
----@param MatchParser Dota2MatchParserInterface
----@param map table
----@return table
-function MapFunctions.getAdditionalExtraData(MatchParser, map)
-	map.extradata.comment = map.comment
-	map.extradata.team1side = MatchParser.getSide(map, 1) or ''
-	map.extradata.team2side = MatchParser.getSide(map, 2) or ''
-	map.extradata.team1objectives = MatchParser.getObjectives(map, 1) or {}
-	map.extradata.team2objectives = MatchParser.getObjectives(map, 2) or {}
-	map.extradata.publisherid = map.publisherid or ''
-
-	return map
+	return extraData
 end
 
 -- Parse participant information
----@param MatchParser Dota2MatchParserInterface
+---@param MatchParser LeagueOfLegendsMatchParserInterface
 ---@param map table
----@param opponents table[]
+---@param opponentCount integer
 ---@return table
-function MapFunctions.getParticipants(MatchParser, map, opponents)
+function MapFunctions.getParticipants(MatchParser, map, opponentCount)
 	local participants = {}
-	local extradata = {}
 	local getCharacterName = FnUtil.curry(MatchGroupInput.getCharacterName, HeroNames)
 
-	for opponentIndex = 1, MAX_NUM_OPPONENTS do
-		local teamPrefix = 'team' .. opponentIndex
-		Array.forEach(MatchParser.getHeroPicks(map, opponentIndex) or {}, function (hero, idx)
-			extradata[teamPrefix .. 'hero' .. idx] = getCharacterName(hero)
-		end)
-		Array.forEach(MatchParser.getHeroBans(map, opponentIndex) or {}, function (hero, idx)
-			extradata[teamPrefix .. 'ban' .. idx] = getCharacterName(hero)
-		end)
+	for opponentIndex = 1, opponentCount do
 		for playerIndex, participant in ipairs(MatchParser.getParticipants(map, opponentIndex) or {}) do
 			participant.character = getCharacterName(participant.character)
 			participants[opponentIndex .. '_' .. playerIndex] = participant
 		end
 	end
```

Comment:
the players auto sorted by role or something like on lol?
if no might attach the char to the wrong player
(additionally i think once substitutes come into play this will attach chars to wrong players)

Reply:
For API matches, the data comes in one go

[function CustomMatchGroupInputMatchPage.getParticipants(map, opponentIndex)](https://github.com/Liquipedia/Lua-Modules/blob/4b23794ea42f31a72318f8eb872b2d5bde671d0b/components/match2/wikis/dota2/match_group_input_custom_match_page.lua#L485-L512)

for manual matches, this will do nothing
[function CustomMatchGroupInputNormal.getParticipants(map, opponentIndex)](https://github.com/Liquipedia/Lua-Modules/blob/4b23794ea42f31a72318f8eb872b2d5bde671d0b/components/match2/wikis/dota2/match_group_input_custom_normal.lua#L30-L32)